### ISSUE TITLE
Graph of obtainment (looking for a name)

### ISSUE THREAD

**Issue by Brambor** [CleverRaven/Cataclysm-DDA#69831]
### Is your feature request related to a problem? Please describe.

I am frustrated when I need an item for crafting and I didn't know I could just disassemble something. Or smash something. Or craft it.

I am also frustrated when I needlessly spend many in-game hours crafting something I should have just disassembled a random TV that is laying around and have it in 1h5m rather than 8 hours.

### Solution you would like.

_Level 1: first implementation_
Graph of obtainment. (wiki for item obtainment)
Input is desired item (or furniture if we don't want just items). It will show all* (see alternatives) ways of obtaining that item, namely:
1. crafting recipes
2. disassembly recipes
3. Cut up things
4. things I can smash (big rock -> small rocks)
5. items I can find (described by spawn pools)
6. construction (constructing furniture)
7. animals laying eggs or producing other produce
8. butchery
9. harvesting crops (need to know about planting crops)
10. cutting trees
11. ? fishing - I never did it in CDDA, is it just kill and butcher with extra steps?
12. ? mining - is it smashing?
13. ??? what did I forget?

_Level 2: recursion_
Show multiple steps: find item X OR cut Y to get item X, then deconstruct item X.
Yes, this is going to be slow, but it is not a background task that needs to be run constantly but rather only every time when the player asks, like search in crafting menu. It provides value despite heavy CPU usage.

**GUI**:
Entry point (after we select desired item) is like Crafting-GUI, where tabs are ways of obtainment (listed above, e.g. crafting, disassembly, butchery), it lists "recipes".
Every "recipe" in that GUI has requirements. If the requirement is an item, we should be able to easily show obtainment graph for that item. It notably has time requirement (mainly used by crafting & disassembly).

_Level 3: consider surroundings_
Like in crafting menu, which shows which recipes can be crafted, highlight obtainment paths that are possible. Additional input is the surrounding the player can see.

Example: Namely cut up curtains in room with windows.

Aditional features:
Save some graph with potentionally selected branches (so that other don't appear and clutter). This would serve as a TODO list for the player - get 7 rags and 6 plastic chunks, then I can craft all the way to this step, obtain a hammer and I can make survival gear.

I would very much like to have a menu, where I can explore the items in game by expanding each node to see all ways it can be obtained, shown in a graphical way that would keep everything. And save that when I am satisfied to come back to it. Tick of individual branches. I did this on paper when playing CDDA.

Time per item:
I have 100 things I hoarded and I don't care about. I don't wanna calculate in my head what I need to disassemble to get most of item X per least of time. I wanna say "obtain copper wire" and it shows me list of ways I can do it with column "time per first item" which I can sort by. Also a column "amortized time per item", for when I need 500 copper wire, not just one.

### Describe alternatives you have considered.

*all ways item can be obtained - this could reveal things to player they shouldn't be revealed, I have two solutions:
1. checkbox that disables this feature, or make it debug menu only.
Or some other, but practically equal way of disabling this feature.

2. Hide everything until player discovered it.
- Crafting recipes are hidden until character sees them for the first time.
- Disassembly recipes are hidden until player sees that item.
- Cut thing up to show
- Smashing is hidden until player smashes that thing (and unblock only the discovered results - only things that dropped?)
- Item find - bit more complicated - lock it until I discover that place with that loot?

Since there are so many ways to obtain GUI would show the 3 or 4 tabs that have "recipes" and one more tab saying "11 more tabs inactive" it can be selected and will show disabled categories, if we want that show somewhere, otherwise it cannot be selected.

### Additional context

I proposed this as my bachelor's work.

This sorta includes taking on the "disassembly crafting menu", or at least exploring the entries it would have.

**Comment by Brambor** [comment]
I am gonna finish writing it latter, class is starting.

**Comment by ZhilkinSerg** [comment]
It sounds a bit like HHG - https://cdda-guide.nornagon.net/

**Comment by CoroNaut** [comment]
I specifically like the to-do list, perhaps there could be a "Gather nearby items" button that could put them into your inventory (only within reach like how far the crafting menu goes to find components). 

I also kinda like the "Item find" thing. My thinking is that it could record for example, how many gun stores you've walked into and can say something like "Contains ammunition, guns" maybe? Then as you visit more gun stores, the loot table gets more refined as you get to 10-20 visited?

When looking at an item, having a specific window to open for showing "you could use it to craft..." and disassembly would be nice. Perhaps a suggestion: It would be nice if the player could select a specific component from the disassembly to prioritize at the expense of losing more of the other components. Currently only item damage limits the components you get from disassembly. Might be best to put that into another feature request issue though, if its worth doing.

The "obtainment graph" might need more explaining or pictures, I'm not sure I follow your plan here exactly. Is it more like a flow chart, and you navigate through a flow chart with the selection cursor? Personally, I like the inventory system that essentially _lists_ items, and has an expansion for items inside containers. Perhaps the listing contains all the items in your inventory, and you can expand each item into its disassembly components? Then as you go through the list and expand more items, you can select items to add them to a to-do list?

**Comment by Brambor** [comment]
> It sounds a bit like HHG - https://cdda-guide.nornagon.net/

Well... ... ... it is kinda exactly like that. At least the traversing. It doesn't contain the todo list aka. obtainment graph.

**Comment by Brambor** [comment]
The obtainment graph:
The current graphical representation in my mind is just a graph
<details> 
  <summary>Google pic of oriented graph</summary>

![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/13402666/f5716266-7a3b-40eb-865c-8cf44573852c)

</details>

Now each node is expendable. Lets say you start with `pair of survivor boots` 
https://cdda-guide.nornagon.net/item/boots_survivor - I am using that now.

- The boots are the only node in a new graph of obtainment.
- You want to obtain the boots, so you expand the boots node.
- The graph expands, adds two nodes
-  - one for(if you discovered that option)
<details>
<summary>SPOILER!</summary>

find it - see the link
</details>

- - one for crafting the boots.
- You can Expand the option of crafting boots.
- Now this node will spawn (around it) a node for each requirement.
- Nodes with items, like `welding_rod_steel` behave the same as boots.
- Node with tools `tool with glare protection of 1 or more.` can be expanded to give tools (or furniture) that satisfy that.


It needs to support AND branches as well as OR branches.

**Comment by Brambor** [comment]
I would also like to get a "list of tools I can craft that have at least one quality better than any tool in my surroundings". But that might be for a different issue/system. It is only a nice-to-have.

**Comment by PatrikLundell** [comment]
I'd suggest the words Acquire/Acquisition.

**Comment by Brambor** [comment]
What I didn't mention - there should be automatic Expansion, which will expand nodes as needed until a solution is found, ideally with items on hand. This might be slow and some sort of heuristic is needed to compare for example looking for an item outside vs crafting an item for a very long time. These decisions are usually for the player to decide, the decision is equivalent to the heuristics. I don't need to find a good heuristic but I need the system to be able to accept one. Possibly, the player could provide simple heuristics, like "don't try to search for rare items, preferably craft them" for some definition of "rare item" based on loot pools.

**Comment by Brambor** [comment]
Note: further down the line, an add-on to the sidebar could be made to track todo.

**Comment by Brambor** [comment]
I will slowly work on it. Don't close it yet, bot.

It will become my bachelor's work. There is still time, thought.

**Comment by PatrikLundell** [comment]
It doesn't matter what the bot does or doesn't. What's important is the merging of PR(s), regardless of how many, if any, suggestions may have preceded them and regardless of their status. The bot just hides the suggestion so it won't be found with normal search profiles.

**Comment by Brambor** [comment]
> It doesn't matter what the bot does or doesn't. What's important is the merging of PR(s), regardless of how many, if any, suggestions may have preceded them and regardless of their status. The bot just hides the suggestion so it won't be found with normal search profiles.

I wanted to make it clear to everyone, that I will work on it, not just bot. But it will take a couple of months.

**Comment by Brambor** [comment]
This is my latest iteration on the idea:

### Usecase
Player wants to craft something, they have an item snapshot (IS) ready as a source.

The following are (can be) displayed in the layout:
- "missing items": negative(total)
- "available items": positive(total)
- "used items": items used for crafting
- "intermediate items": what player will craft
- "recipes": what recipes in what amount are to be executed
- "want": items player wants, doesn't include intermediate recipes
- "time": estimated total time needed to finish all crafts
"total items" (not shown): IS - want + intermediate - used

positive(IS) - shows only items from IS that have positive count
negative(IS) - shows only items from IS that have negative count

Want phase:
1. Player clicks what they want to have - this adds item to "want".

Equalize phase:
1. They select crafting recipe (and count) and what will they use to craft it.
	- add recipe to recipes. recipes can be changed / added / removed
	- add result to intermediate
	- add used sources to used
	- add total time
3. If there are no missing items, the plan is marked as ready, player can go back to step 1. or save the plan.
4. QoL if there are any unneeded recipes, warn player, player can remove them.
5. Find a craft order that would work.

**Comment by IdleSol** [comment]
Determine in advance how your graph will respond to feedback loops. Just an example, creating a hammer with quality 3 requires a tool with hammering quality 3. And this is either the hammer itself or its bronze version.

For more complex examples, you can look here: https://github.com/CleverRaven/Cataclysm-DDA/issues/72208#issuecomment-2007075330

**Comment by Brambor** [comment]
Thanks for the example! I want to cut branches of the graph for some loops.

Tools example
1. obtain hammer > obtain by crafting hammer > need hammering 3
2. **craft** hammering 3 tool: don't allow any recipe that needs hammering >= 3; allow **finding** any hammer
3. if we allowed bronze hammer, to craft it, now we cannot allow any recipe requiring hammering >= 3 in this branch.

Component example, I want to find loops where we get C from crafting with recipe R recursively getting back to requiring component C. If this loop doesn't produce any extra C, then don't include this loop at all.
There is some water example:
1. to obtain water
2. make potassium hydroxide which has 1 water as a byproduct
3. use 3 clean water to use it
4. craft 1 clean water from 1 water

This is a negative loop, but stable loops are not acceptable either.

For automatic Node expansion loops definitely should be solved. For manual Node expansion it is a QoL, but I definitely want to solve it.

**Comment by Brambor** [comment]
I had issues with snippets, Renech solved similar issues here https://discord.com/channels/598523535169945603/598529174302490644/1251822245853270086 (note to self).

**Comment by Brambor** [comment]
"Goals" sounds better than "Acquire graph"
